R/namespace.R

```diff
@@ -1,15 +1,71 @@
-# Parse namespace files and return imports exports, methods
+#' Extract `NAMESPACE` imports data frame from package path
+#'
+#' @keywords internal
+#' @noRd
 namespace_imports <- function(path = find_package()) {
-  namespace_data <- tryCatch(
-    parseNamespaceFile(basename(path), package.lib = file.path(path, "..")),
+  pkg <- tryCatch(
+    basename(path),
     error = function(e) NULL
   )
 
-  if (length(namespace_data$imports) == 0L) {
+  extract_namespace_imports_df(pkg)
+}
+
+#' Extract `NAMESPACE` imports into a data frame
+#'
+#' `getNamespaceImports()` makes sure that `except` directives are respected.
+#'
+#' @examples
+#' lintr:::extract_namespace_df()
+#' lintr:::extract_namespace_df("base")
+#' lintr:::extract_namespace_df("rex")
+#'
+#' @keywords internal
+#' @noRd
+extract_namespace_imports_df <- function(pkg = NULL) {
+  if (!is.null(pkg)) {
+    ns_imports <- tryCatch(
+      getNamespaceImports(pkg),
+      error = function(e) NULL # if an imported namespace is unavailable
+    )
+  }
+
+  if (is.null(pkg) || length(ns_imports) == 0L) {
     return(empty_namespace_data())
   }
 
-  do.call(rbind, lapply(namespace_data$imports, safe_get_exports))
+  ns_imports[["base"]] <- NULL
+  pkg_names <- names(ns_imports)
+  df <- do.call(
+    rbind.data.frame,
+    c(
+      mapply(ns_imports_list_to_df, ns_imports, pkg_names, SIMPLIFY = FALSE),
+      stringsAsFactors = FALSE
+    )
+  )
```

Comment:
I am fairly certain that the test failures on `R < 4.0` are due to `stringsAsFactors` not being set to `FALSE` here. Something not right the way I am passing `stringsAsFactors` to `rbind()` or `rbind.data.frame()` via `do.call()`.